web/pandas/pdeps/0006-ban-upcasting.md

```diff
@@ -0,0 +1,178 @@
+However, this can potentially hide bugs, break user expectations, and copy data
+in what looks like it should be an inplace operation.
+
+An example of it hiding a bug is:
+```python
+In[9]: ser = pd.Series(pd.date_range("2000", periods=3))
+
+In[10]: ser[2] = "2000-01-04"  # works, is converted to datetime64
+
+In[11]: ser[2] = "2000-01-04x"  # typo - but pandas does not error, it upcasts to object
+```
+
+The scope of this PDEP is limited to setitem-like operations on Series.
+For example, starting with
+```python
+df = DataFrame({"a": [1, 2, np.nan], "b": [4, 5, 6]})
+ser = df["a"].copy()
+```
+then the following would all raise:
+- ``ser[0] = 'foo'``;
+- ``ser.fillna('foo', inplace=True)``;
+- ``ser.where(ser.isna(), 'foo', inplace=True)``
+- ``ser.fillna('foo', inplace=False)``;
+- ``ser.where(ser.isna(), 'foo', inplace=False)``
+- ``ser.iloc[0] = 'foo'``
+- ``ser.loc[0] = 'foo'``
+- ``df.loc[indexer, 'a'] = 'foo'``
+- ``ser[indexer] = 'foo'``
+
+where ``indexer`` could be a slice, a mask, a single value, a list or array of values,
+or any other allowed indexer.
+
+Examples of operations which would not raise are:
+- ``ser.diff()``;
+- ``pd.concat([ser, ser.astype(object)])``;
+- ``ser.mean()``;
+- ``ser[0] = 3.``;
+- ``ser[0] = 3``.
+
+## Detailed description
+
+Concretely, the suggestion is:
+- if a ``Series`` is of a given dtype, then a ``setitem``-like operation should not change its dtype;
+- if a ``setitem``-like operation would previously have changed a ``Series``' dtype, it would now raise.
+
+For a start, this would involve:
+
+1. changing ``Block.setitem`` such that it does not have an ``except`` block in
+
+  ```python
+  value = extract_array(value, extract_numpy=True)
+  try:
+      casted = np_can_hold_element(values.dtype, value)
+  except LossySetitemError:
+      # current dtype cannot store value, coerce to common dtype
+      nb = self.coerce_to_target_dtype(value)
+      return nb.setitem(indexer, value)
+  else:
+  ```
+
+2. making a similar change in:
+   - ``Block.where``;
+   - ``Block.putmask``;
+   - ``EABackedBlock.setitem``;
+   - ``EABackedBlock.where``;
+   - ``EABackedBlock.putmask``;
+   - ``_iLocIndexer._setitem_single_column``;
+
+The above would already require several hundreds of tests to be adjusted.
+
+### Ban upcasting altogether, or just upcasting to ``object``?
+
+The trickiest part of this proposal concerns what to do when setting a float in an integer column:
+
+```python
+In[1]: ser = pd.Series([1, 2, 3])
+
+In[2]: ser[0] = 1.5
+```
+
+This is not necessarily a sign of a bug, because the user might just be thinking of their ``Series`` as being
+numeric (without much regard for ``int`` vs ``float``) - ``'int64'`` is just what pandas happened to infer.
+
+Possible options could be:
+1. just raise, forcing users to be explicit;
+2. convert the float value to ``int`` before setting it;
+3. limit "banning upcasting" to when the upcasted dtype is ``object``.
```

Comment:
we already do convert 1.0 to 1

Reply:
yes, but this proposal would disallow that

I hadn't considered that it could only be allowed in round cases...I'll think about it and revisit the proposal

Reply:
> yes, but this proposal would disallow that

Sorry, my bad - `1.0` would be compatible with the Series' dtype, and so we could get
```python
In [8]: ser = pd.Series([1,2,3])

In [9]: ser
Out[9]:
0    1
1    2
2    3
dtype: int64

In [10]: ser[0] = 4.0

In [11]: ser
Out[11]:
0    4
1    2
2    3
dtype: int64
```
(just like we currently do, for both `'int64'` and `'Int64'`)

Still, I'll add this as an example to clarify that it wouldn't change

Reply:
Actually, this is already given as an example here:

https://github.com/pandas-dev/pandas/blob/875cf4c53313a8b0aca72766c25198b35622314c/web/pandas/pdeps/0006-ban-upcasting.md#L73-L78

Reply:
That example (`ser[0] = 3.`) uses a "round" float, so I would say that is the easy case (we already preserve the dtype in those cases, so we can probably keep doing that?). But the example in the section here uses a float that would be truncated if we preserve the dtype. The current behaviour of that example:

```
>>> ser = pd.Series([1, 2, 3])
>>> ser
0    1
1    2
2    3
dtype: int64
>>> ser[0] = 1.5
>>> ser
0    1.5
1    2.0
2    3.0
dtype: float64
```

Maybe it would be good to show the output in the text, to make it clear what the current behaviour is. We should probably also explicitly mention those two different cases, and then assume that the rest of the text is about the case where the float would get truncated? 

Because when just looking that the options as they are listed here now, I think the current idea is a mix of options 1 and 2? I.e. convert float to int when possible, otherwise raise (in which case the user needs to be explicit and `astype` to float first)

Reply:
thanks, have added the output